File: destack-py/destack/store/memory/entity/wiring.py
from destack.language import (
    Entity,
    Materialization,
    Node,
    NodeReference,
    NodeType,
    Value,
)
from destack.language.registry import NODE_TYPE_SCALAR_BY_NODE_TYPE
from destack.utils.uuid import UUID

from .core import MemoryEntityRow

NODE_METATYPE_KEY = str(Node.property("metatype").id)
NODE_ID_KEY = str(Node.property("id").id)
NODE_PARENT_PTR_KEY = str(Entity.property("parent").id)
NODE_SPACE_PTR_KEY = str(Node.property("space").id)
ENTITY_DEFINITION_PTR_KEY = str(Entity.property("definition").id)

ENTITY_SNAPSHOT_PTR_KEY = str(Entity.property("snapshot").id)
ENTITY_MATERIALIZATION_KEY = str(Entity.property("materialization").id)

NODE_REFERENCE_ID_KEY = str(NodeReference.property("id").id)


def pack_entity_row(value: Value) -> MemoryEntityRow:
    """Pack a Value into a MemoryRow."""
    value_cson = value.value
    assert value_cson is not None, f"no value for {value!r}"
    node_type = NodeType(value_cson[NODE_METATYPE_KEY])
    id = UUID(value_cson[NODE_ID_KEY])
    ptr = NodeReference(
        type=node_type,
        id=UUID(value_cson[NODE_ID_KEY]),
        space_id=UUID(value_cson[NODE_SPACE_PTR_KEY][NODE_REFERENCE_ID_KEY])
        if NODE_SPACE_PTR_KEY in value_cson
        else None,
        definition_id=UUID(value_cson[ENTITY_DEFINITION_PTR_KEY][NODE_REFERENCE_ID_KEY])
        if ENTITY_DEFINITION_PTR_KEY in value_cson
        else None,
    )
    parent_ptr = value_cson.get(NODE_PARENT_PTR_KEY)
    if parent_ptr is not None:
        parent_ptr = NodeReference.from_cson(parent_ptr)
    snapshot_ptr = value_cson.get(ENTITY_SNAPSHOT_PTR_KEY)
    if snapshot_ptr is not None:
        snapshot_ptr = NodeReference.from_cson(snapshot_ptr)
    materialization = value_cson.get(ENTITY_MATERIALIZATION_KEY)
    if materialization is not None:
        materialization = Materialization(materialization)
    row = MemoryEntityRow(
        metatype=node_type,
        id=id,
        snapshot_id=snapshot_ptr.id if snapshot_ptr is not None else None,
        materialization=materialization,
        ptr=ptr,
        parent_ptr=parent_ptr,
        value=value_cson,
    )
    return row


def unpack_entity_row(row: MemoryEntityRow) -> Value:
    """Unpack a MemoryRow to a Value."""
    type = NODE_TYPE_SCALAR_BY_NODE_TYPE[row.metatype]
    return Value(type=type, value=row.value)

File: destack-ts/src/store/memory/entity/wiring.ts
import {
  Materialization,
  NODE_TYPE_SCALAR_BY_TYPE,
  NodeReference,
  NodeType,
  Value,
} from "@destack/language";
import {
  ENTITY_MATERIALIZATION_KEY,
  ENTITY_SNAPSHOT_KEY,
  ENTITY_DEFINITION_PTR_KEY,
  NODE_ID_KEY,
  NODE_METATYPE_KEY,
  ENTITY_PARENT_KEY,
  NODE_REFERENCE_ID_KEY,
  NODE_SPACE_PTR_ID,
} from "@destack/store/memory/core";
import { MemoryEntityRow } from "@destack/store/memory/entity/core";

/**
 * Pack a Value into a MemoryEntityRow.
 */
export function packEntityRow(value: Value): MemoryEntityRow {
  const valuePacked = value.value;
  if (!valuePacked) {
    throw new Error(`no value for ${value.repr()}`);
  }

  const nodeType = Number(valuePacked[NODE_METATYPE_KEY]) as NodeType;
  const id = String(valuePacked[NODE_ID_KEY]);
  const ptr = new NodeReference({
    type: nodeType,
    id: String(valuePacked[NODE_ID_KEY]),
    spaceId:
      NODE_SPACE_PTR_ID in valuePacked
        ? String(valuePacked[NODE_SPACE_PTR_ID][NODE_REFERENCE_ID_KEY])
        : undefined,
    definitionId:
      ENTITY_DEFINITION_PTR_KEY in valuePacked
        ? String(valuePacked[ENTITY_DEFINITION_PTR_KEY][NODE_REFERENCE_ID_KEY])
        : undefined,
  });

  let parentPtr: NodeReference | null = null;
  const parentPtrValue = valuePacked[ENTITY_PARENT_KEY];
  if (parentPtrValue !== undefined) {
    parentPtr = NodeReference.fromValue(parentPtrValue);
  }

  let snapshotPtr: NodeReference | null = null;
  const snapshotPtrValue = valuePacked[ENTITY_SNAPSHOT_KEY];
  if (snapshotPtrValue !== undefined) {
    snapshotPtr = NodeReference.fromValue(snapshotPtrValue);
  }

  let materialization: Materialization | null = null;
  const materializationValue = valuePacked[ENTITY_MATERIALIZATION_KEY];
  if (materializationValue !== undefined) {
    materialization = materializationValue as Materialization;
  }

  const row = new MemoryEntityRow({
    metatype: nodeType,
    id,
    snapshotId: snapshotPtr?.id || null,
    materialization,
    ptr,
    parentPtr,
    value: valuePacked,
  });

  return row;
}

/**
 * Unpack a MemoryEntityRow to a Value.
 */
export function unpackEntityRow(row: MemoryEntityRow): Value {
  const type = NODE_TYPE_SCALAR_BY_TYPE[row.metatype];
  return new Value({ type, value: row.value });
}
